### DISCUSSION TITLE
Ability to upload local image to photoprism

### DISCUSSION THREAD

**Discussion by o0Zz** [Radiokot/photoprism-android-client#58] [Ideas]
Hi, 

I'm interressted to have an equivalent to Google photos, meaning be able to see my photoprism library but also to automatically detect new pictures on the phone and upload them in background without any user interaction (With some options like, when wifi is available etc...).

Do you have any plan for such kind of feature ?
I can contribute to develop it and send a pull request, however I want to make sure first, it make sense for you :)

I found some application on the playstore to sync with photoprism or webdav but none of them match my needs (Either not working, or need to pay, or with ads, or no automatic upload)

Thank you and great work by the way ;)

**Comment by Radiokot** [comment]
Hi. Thank you for checking out the app.

The gallery is not intended to sync content with the library. Sync is a separate, complicated process with lots of edge-cases and workarounds, where error causes a mess in a library. I don't want to bother with it, especially when there are well tested apps doing the thing. 

I recommend using [Autosync](https://play.google.com/store/apps/details?id=com.ttxapps.autosync). It has many supported sync methods, including WebDAV. For each method, they give a free trial, so you can test it for your use-case. Lifetime license is cheap. Me with my wife use this app to sync our photos in the upload mirror mode, it works flawlessly.

**Comment by o0Zz** [comment]
Yep I do agree, It's better to have 1 app that does 1 thing well than an app that does 1000 things badly.

You're right, Autosync is one of the best one, however they don't have a trigger "On new picture". Meaning you have to manually select the folder you want to sync. So. if you receive a photo on what's app for example, you have to manually select the what's app download folder. It's not really convenient when you have multiple applications and multiple phones.

I will see how to deal with that, but I'm really motivated to do a fork of your app for that ^^.

**Reply by Radiokot** [reply]
I really recommend you first try to do this with the existing tools.
Autosync has an [automation API](https://metactrl.com/autosync/automation/) which allows other apps triggering the sync. On the other hand, there are several powerful automation apps that can use the API and have file system triggers. For example, this video shows how to trigger a Tasker task when a new photo appears: https://www.youtube.com/watch?v=HOJABTNgf04

Of course, if you see this as a challenge to take on your own, program it! 💪🏻

**Comment by ippocratis** [comment]
There is also foldersync,  another paid sync app that is also great
The free version has all features available but with adds
It supports many cloud providers for sync and many sync protocols
bidirectional or one directional sync is available ,WebDAV mtls supported and most important auto sync is available 

https://foldersync.io/docs/help/quickstart/#configure-folderpair

For multiple photo folder imao dcim and pictures folders will cover 99% of the apps that store images/videos on the device 

I will definitely try the fork though if take the fork route 

:-)

**Comment by MiSebCo** [comment]
Would it be possible to have the app as a target that images can be shared to (e.g. from the album app to the gallery app)? As a simple way to upload an image to photoprism (no sync) without any further checks. Thank you.

**Reply by Radiokot** [reply]
Simple import by sharing files to the app  (like Upload button in the web app) looks like a nice feature to have. I'll add it to the backlog. Thank you for suggestion, Michael.

**Reply by Radiokot** [reply]
Working on it.

**Reply by Radiokot** [reply]
Import through sharing is added in 1.28.0. Open the app once after the installation so the component gets enabled
https://github.com/Radiokot/photoprism-android-client/releases/tag/1.28.0(42)

**Reply by MiSebCo** [reply]
This is great! Like it how you solved getting the additional permissions for the app during the first usage of the feature and the clear explanations to the user. 

However: I always get an error when using the feature. Tried it on 2 devices. Is there any error log I could check?

I am using Photoprism Build [240711-2197af848]. 

What I tried so far: 
* Tried an oder version of Photoprism 
* Reconnected Gallery App to the Photoprism Library

Let me know what additional information I can provide.

**Reply by Radiokot** [reply]
Can you please try connecting to https://demo.photoprism.app and uploading something there?

**Reply by MiSebCo** [reply]
This worked. I downloaded the picture from Oct 7 2018 from the demo server and uploaded it again via the sharing feature. The app confirmed success - I could not see the uploaded picture in the library, however. Maybe because it already was in the library.

**Reply by Radiokot** [reply]
Thanks for checking it, Michael. It seems that upload fails with password-protected libraries. Will publish a fix soon.

**Reply by Radiokot** [reply]
Here it is:
https://github.com/Radiokot/photoprism-android-client/releases/tag/1.28.1(43)
Sorry for the inconvenience, it was an obvious mistake 🙄

**Reply by MiSebCo** [reply]
Yes - it works now. Great. Thank you! This makes uploading so simple and intuitive - especially for people being used to this workflow from other apps. Perfect.

**Comment by jrstrunk** [comment]
If anyone does want to take on the challenge of programming a native photo sync functionality, the NextCloud Android app has sync implemented well and is open source. Here is a place you could get started: https://github.com/nextcloud/android/blob/master/app/src/main/java/com/nextcloud/client/jobs/FilesSyncWork.kt

**Comment by neon12345** [comment]
I have a simple upload and sync client/server implementation [here](https://codeberg.org/Neon12345/file_share_server).
- only host, port, password and optional sync folder options
- aes encryption
- each request creates a new unique root directory => no override on upload
- server sha256 file cache to prevent duplicate uploads
- share files with the app or upload the sync directory.

I do not plan to release this as an official app myself.

Edit: The idea behind this approach is to make the upload "append data only" and manage/delete externally (e.g. photoprism web frontend).  Host, port and password are the only information needed for uploads, which requires cooperation from the server but has less room for user errors.

**Comment by master3395** [comment]
I made an Android app, not released anywhere yet.
NT Prism is what I've called it so far.

It supports:
1. Gallery(Library) from Photoprism.
2. Manual Upload and auto upload each 15 minutes (Google minimum requirements)
3. Choose a local folder or folders.
4. Upload from WebDAV, SMB, Google Drive, OneDrive, Dropbox, and Mega (requires your own OAuth ID)
5. Both image and video upload.
6. It supports Cloudflare Tunnel and  Direct Lan/IP and Tailscale.
7. It supports a fallback if the first option does not work.
8. It supports HTTP Basic Auth.
9. It supports insecure SSL
10. It supports delete after upload
11. mTLS
12. It supports Photos only in settings
13. It supports Wi-fi only
14. It supports only while charging.
15. It supports showing notifications on Android.

<img width="1440" height="3216" alt="image" src="https://github.com/user-attachments/assets/45022380-479a-490e-b77e-01b68aa1f475" />
<img width="1440" height="3216" alt="image" src="https://github.com/user-attachments/assets/c4e291f8-c109-4ac0-83a4-23743b47557e" />
<img width="1440" height="3216" alt="image" src="https://github.com/user-attachments/assets/1006db68-fff1-4ecb-8c93-548fecf0d908" />
<img width="1440" height="3216" alt="image" src="https://github.com/user-attachments/assets/c700f649-fe17-4c20-adb8-a6f54e5407b8" />

**Reply by Radiokot** [reply]
Hey, it's cool, publish it on GitHub :)
